💃 🆕-style models

cthoyt opened the pull request:
This PR is nearing the end of the #107 demolition train following ~#257~, ~#266~, ~#274~, ~#277~, ~#287~, and ~#292~. It does the following:

1. Implements `pykeen/models/nbase.py` which holds the new abstract model and a few utility functions to support it.
2. Adds an additional canonical form to the `pykeen.nn.Embedding` class - this differs from old-style models to new-style models
3. As an example, updates Unstructured Model; a49fe48d

In future PRs, other models can be ported individually. This allows us to keep the performance benefits of old-style models (in the few cases where it exists) while also giving the flexibility of new-style models. One exceptional example is that R-GCN can be re-implemented much more elegantly with new-style models. We also need to make a new tutorial on how to write your own model, but that's been partially written in #107.

cthoyt committed:
Step 1: minimal example

cthoyt committed:
Update hierarchy

cthoyt committed:
Update representation.py

cthoyt committed:
Ready for testing

Trigger CI

cthoyt committed:
Cleanup

Trigger CI

cthoyt commented on src/pykeen/models/nmodel/base.py:
```diff
@@ -0,0 +1,547 @@
+        ).view(ht_batch.shape[0], self.num_relations)
+
+
+def collect_regularization_terms(module: nn.Module) -> Union[float, torch.FloatTensor]:
```
This won't work because the regularization class hasn't been updated. I guess we have to take the entire new set of regularizers over for this

mberr replied:
@cthoyt this comment is fixed by #266 ?

cthoyt replied:
not yet. i'm actually looking at this code ATM

cthoyt committed:
Add new style regularizer

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Update trans_h.py

cthoyt committed:
Update utilities

cthoyt committed:
Update trans_h.py

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Update cases.py

Trigger CI

cthoyt committed:
Remove deprecated loss stuff

cthoyt committed:
Align regularizers

cthoyt commented on src/pykeen/models/nmodel/regularizers.py:
```diff
@@ -0,0 +1,256 @@
+_REGULARIZER_SUFFIX = 'Regularizer'
+
+
+class NewRegularizer(Regularizer):
```
The creation of new-style regularizers should probably get split into its own PR as well, since I made a few changes to the base class to facilitate this

cthoyt committed:
cleanup

cthoyt commented on src/pykeen/regularizers.py:
```diff
@@ -77,18 +77,26 @@ def forward(self, x: torch.FloatTensor) -> torch.FloatTensor:
-    def update(self, *tensors: torch.FloatTensor) -> None:
+    def update(self, *tensors: torch.FloatTensor) -> bool:
         """Update the regularization term based on passed tensors."""
+        # TODO if not self.training or not torch.is_grad_enabled() or (self.apply_only_once and self.updated):
```
@mberr we changed this part of the update step in the new code, should we also apply it to the old code? (the checking if in training mode and that grad is enabled)

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Update regularizers

cthoyt committed:
Update TransH regularizer

cthoyt committed:
Reorganize

cthoyt committed:
Finish reorg

Trigger CI

mberr commented on src/pykeen/models/nemb.py:
```diff
@@ -0,0 +1,275 @@
+        d_i = 1 else. If indices is not None, then batch_size = indices.shape[0], and d_i = 1 if
+        indices.ndimension() = 1 else d_i = indices.shape[1]
+
+        The canonical shape is given by (batch_size, 1, ``*``) if indices is not None, where batch_size=len(indices),
```
this docstring contradicts the one from above.

mberr replied:
The one from before seems to be correct, and this one erroneous

mberr commented on src/pykeen/utils.py:
```diff
@@ -917,6 +930,43 @@ def extend_batch(
+    dims: Dict[str, Tuple[int, ...]] = dict()
+    errors = []
+    for actual_shape, shape in x:
+        if isinstance(actual_shape, torch.Tensor):
```
use `torch.is_tensor`?

cthoyt replied:
nope, `torch.is_tensor()` breaks mypy (this code is exactly what's in torch.is_tensor though)

mberr commented on src/pykeen/utils.py:
```diff
@@ -917,6 +930,43 @@ def extend_batch(
+) -> bool:
+    """Verify that a sequence of tensors are of matching shapes.
+
+    :param x:
```
from the type annotation / code below it looks like the first element of the tuple can also be a tensor's shape, i.e. a tuple of integers.

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+        """Instantiate the regularizer from this class's default settings.
+
+        If the default regularizer is None, None is returned.
+        Handles the corner case when the default regularizer's keyword arguments are None
```
Missing `.` in docstring?

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+        _kwargs.update(kwargs)
+        return self.regularizer_default(**_kwargs)
+
+    def post_parameter_update(self) -> None:
```
This is called in the training loops. Hence each `Model` needs to implement this method. Should we move it to the base class?

cthoyt replied:
Good call. Done in 7d4a4b59

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+    ) -> torch.FloatTensor:
+        """Compute the loss for functions requiring two separate tensors as input.
+
+        :param tensor_1: shape: s
```
shape only needs to be broadcastable, not exactly the same

mberr replied:
done in ad58b4dfa2e89d556ee1e6bbd122220ea91ae8f1

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+        This method takes head, relation and tail indices and calculates the corresponding score.
+
+        .. note ::
+            All indices which are not None, have to be either 1-element or have the same shape, which is the batch size.
```
from the `Embedding`'s `forward` description, we also allow 
* None
* 1-d, shape: `(batch_size,)`
* 2-d, shape: `(batch_size, num_heads)` / `(batch_size, num_relations)` / `(batch_size, num_tails)`

cthoyt replied:
@mberr would you please update the docstring to look the way you'd like? Thanks :)

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+
+        :param triples_factory:
+            The triples factory facilitates access to the dataset.
+        :param loss:
```
missing parameters: interaction, entity_representations, relation_representations

cthoyt replied:
Done in 50245c60 (also added darglint)

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+        :return: shape: (batch_size, num_heads, num_relations, num_tails)
+            The score for each triple.
+        """
+        h, r, t = self._get_representations(h_indices, r_indices, t_indices)
```
Pass by keyword instead of positional? Makes it less prone to errors when changing parameter order.

cthoyt replied:
Done in 50245c60

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+        """
+        h, r, t = self._get_representations(h_indices, r_indices, t_indices)
+        scores = self.interaction.score(h=h, r=r, t=t, slice_size=slice_size, slice_dim=slice_dim)
+        return self._repeat_scores_if_necessary(scores, h_indices, r_indices, t_indices)
```
Same here: pass by keyword?

cthoyt replied:
Done in 50245c60

mberr commented on src/pykeen/models/nbase.py:
```diff
@@ -0,0 +1,475 @@
+        h_indices: Optional[torch.LongTensor],
+        r_indices: Optional[torch.LongTensor],
+        t_indices: Optional[torch.LongTensor],
+    ) -> torch.FloatTensor:
```
Add a small docstring?

mberr replied:
done in d575082

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Update representation module

cthoyt committed:
Add complex datatype annotations

Trigger CI

cthoyt committed:
Update complex.py

Trigger CI

cthoyt committed:
Merge branch 'complex-shapes-norman-bates' into new-model-on-the-block

cthoyt committed:
Update embedding specification

cthoyt committed:
Backwards compatible property

cthoyt committed:
Add alternate MORE canonical dimension getter

cthoyt committed:
Update docs

Trigger CI

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Update trans_h.py

cthoyt committed:
Add missing base function

Trigger CI

cthoyt committed:
Revert TransH

cthoyt committed:
Update trans_r.py

cthoyt committed:
Update trans_h.py

cthoyt committed:
Remove unused function

This is for TransH
Trigger CI

cthoyt committed:
Update test_models.py

Trigger CI

cthoyt committed:
Update test_models.py

Trigger CI

mberr commented on src/pykeen/models/base.py:
```diff
@@ -170,6 +170,10 @@ def _reset_parameters_(self):  # noqa: D401
         """Reset all parameters of the model in-place."""
         raise NotImplementedError
 
+    @abstractmethod
```
Do we want to enforce that each model implements this method, even if most models may not enforce any constraints?

cthoyt replied:
You mean leave as an empty function instead of a proper abstract function? Or have a check inside the model to see if it's been implemented?

cthoyt replied:
I see what you mean. Addressed in 5a3ff13d

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Remove abstraction

@mberr you're right - models shouldn't be forced to implement this if they don't want

Trigger CI

cthoyt commented on src/pykeen/models/unimodal/unstructured_model.py:
```diff
@@ -51,40 +48,26 @@ def __init__(
         :param embedding_dim: The entity embedding dimension $d$. Is usually $d \in [50, 300]$.
         :param scoring_fct_norm: The $l_p$ norm. Usually 1 for UM.
         """
+        self.embedding_dim = embedding_dim
```
is this necessary?

cthoyt replied:
I guessed it was for testing purposes, but I managed to remove it and nothing started failing

cthoyt committed:
Update unstructured_model.py

mberr committed:
update docstring of check_shapes

mberr committed:
Update note for new-style model forward

mberr committed:
Add small docstrings to private method

mberr committed:
Merge branch 'master' into new-model-on-the-block

PyKEEN-bot committed:
Trigger CI

mberr committed:
remove trailing whitespace

Trigger CI

mberr committed:
Merge remote-tracking branch 'origin/new-model-on-the-block' into new-model-on-the-block

PyKEEN-bot committed:
Trigger CI

mali-git commented on src/pykeen/nn/emb.py:
```diff
@@ -118,6 +118,47 @@ def get_in_canonical_shape(
+
+        fulfilling the following properties:
+
+        Let i = dim. If indices is None, the return shape is (1, d_1, d_2, d_3) with d_i = num_representations,
```
Maybe add a concrete example?

cthoyt replied:
@mberr it wouldn't hurt to have some unit tests for this, but I never found any in #107 to port here

mberr replied:
> Maybe add a concrete example?

done in 84648c9441b58834aec8b07c53a19d0a7efea5e2

mberr replied:
> @mberr it wouldn't hurt to have some unit tests for this, but I never found any in #107 to port here

done in 982df01c0d2bddca8fe92bb6fb105ecb3d738c62

mali-git commented on src/pykeen/utils.py:
```diff
@@ -931,6 +933,43 @@ def extend_batch(
+        A tuple (t, s), where `t` is a tensor, or an actual shape of a tensor (a tuple of integers), and `s` is a
+        string, where each character corresponds to a (named) dimension. If the shapes of different tensors share a
+        character, the corresponding dimensions are expected to be of equal size.
+    :param raise_on_errors:
```
Maybe add a concrete example?

mberr replied:
done in 22bf857177d3ea03f51c4e325e78a5e84e604d8f

mberr committed:
Add concrete example to check_shapes

mberr committed:
Add examples for get_in_more_canonical_shape

mberr committed:
Add tests for more canonical shape

mberr committed:
Add note to compute_loss' docstring

cthoyt committed:
Merge branch 'master' into new-model-on-the-block

cthoyt committed:
Create extending_models_new.rst

Trigger CI

cthoyt committed:
Update extending_models_new.rst

Trigger CI

cthoyt committed:
Update utils.py

cthoyt committed:
Turn on doctest

Trigger CI

cthoyt committed:
Update utils.py

Trigger CI
